core/src/test/java/io/grpc/internal/AbstractServerStreamTest.java

```diff
@@ -0,0 +1,335 @@
+
+  @Rule public final ExpectedException thrown = ExpectedException.none();
+
+  private final WritableBufferAllocator allocator = new WritableBufferAllocator() {
+    @Override
+    public WritableBuffer allocate(int capacityHint) {
+      return new ByteWritableBuffer(capacityHint);
+    }
+  };
+
+  @Test
+  public void setListener_setOnlyOnce() {
+    AbstractServerStreamBase stream = new AbstractServerStreamBase(allocator, MAX_MESSAGE_SIZE);
+    stream.setListener(new ServerStreamListenerBase());
+    thrown.expect(IllegalStateException.class);
+
+    stream.setListener(new ServerStreamListenerBase());
+  }
+
+  @Test
+  public void setListener_readyCalled() {
+    AbstractServerStreamBase stream = new AbstractServerStreamBase(allocator, MAX_MESSAGE_SIZE);
+
+    ServerStreamListener streamListener = mock(ServerStreamListener.class);
+    stream.setListener(streamListener);
+
+    verify(streamListener).onReady();
+  }
+
+  @Test
+  public void setListener_failsOnNull() {
+    AbstractServerStreamBase stream = new AbstractServerStreamBase(allocator, MAX_MESSAGE_SIZE);
+    thrown.expect(NullPointerException.class);
+
+    stream.setListener(null);
+  }
+
+  @Test
+  public void receiveMessage_listenerCalled() {
+    final ServerStreamListener streamListener = mock(ServerStreamListener.class);
+    AbstractServerStreamBase stream = new AbstractServerStreamBase(allocator, MAX_MESSAGE_SIZE) {
+      @Override
+      protected ServerStreamListener listener() {
+        return streamListener;
+      }
+    };
+
+    // Normally called by a deframe event.
+    stream.receiveMessage(new ByteArrayInputStream(new byte[]{}));
+
+    verify(streamListener).messageRead(isA(InputStream.class));
+  }
+
+  @Test
+  public void receiveMessage_failsIfHalfClosed() {
+    AbstractServerStreamBase stream = new AbstractServerStreamBase(allocator, MAX_MESSAGE_SIZE);
+    // Simulate being closed, without invoking the listener
+    stream.inboundPhase(Phase.STATUS);
+
+    thrown.expect(IllegalStateException.class);
+
+    // Normally called by a deframe event.
+    stream.receiveMessage(new ByteArrayInputStream(new byte[]{}));
+  }
+
+  @Test
+  public void writeHeaders_failsOnNullHeaders() {
+    AbstractServerStreamBase stream = new AbstractServerStreamBase(allocator, MAX_MESSAGE_SIZE);
+    thrown.expect(NullPointerException.class);
+
+    stream.writeHeaders(null);
+  }
+
+  @Test
+  public void writeHeaders_failsIfAlreadySent() {
+    AbstractServerStreamBase stream = new AbstractServerStreamBase(allocator, MAX_MESSAGE_SIZE);
+    stream.writeHeaders(new Metadata());
+    thrown.expect(IllegalStateException.class);
+
+    stream.writeHeaders(new Metadata());
+  }
+
+  @Test
+  public void writeHeaders() {
+    final AtomicReference<Metadata> capturedHeaders = new AtomicReference<Metadata>(null);
+    Metadata headers = new Metadata();
+    AbstractServerStreamBase stream = new AbstractServerStreamBase(allocator, MAX_MESSAGE_SIZE) {
```

Comment:
At some point I kind of expected AbstractServerStreamBase to be split off (like AbstractClientStreamBase).  I think this version of the test is more readable, since everything you need to know about the test is contained in the test body.  Moving the headers to AbstractServerStreamBase would mean that every test has access to them.  If multiple tests wanted to do something with headers, will capturedHeaders be the variable to use?  

Similar feelings towards the other comments below.